cmd/vela-email/plugin.go

```diff
@@ -0,0 +1,282 @@
+	ErrorEmptyAttach = errors.New("attachment provided is empty")
+
+	// ErrorMissingSMTPParam is returned when the plugin is missing a smtp host or port parameter.
+	ErrorMissingSMTPParam = errors.New("missing smtp parameter (host/port)")
+
+	// ErrorMissingSMTPUsernameParam is returned when the plugin is missing smtp username parameter.
+	ErrorMissingSMTPUsernameParam = errors.New("missing smtp username")
+
+	// ErrorMissingSMTPPasswordParam is returned when the plugin is missing smtp password parameter.
+	ErrorMissingSMTPPasswordParam = errors.New("missing smtp password")
+)
+
+// Plugin represents the configuration loaded for the plugin.
+type (
+	Plugin struct {
+		// Email arguments loaded for the plugin
+		Email *email.Email
+		// Attachment arguments loaded for the plugin
+		Attachment *email.Attachment
+		// SmtpHost arguments loaded for the plugin
+		SMTPHost *SMTPHost
+		// TLSConfig arguments loaded for the plugin
+		TLSConfig *tls.Config
+		// SendType arguments loaded for the plugin
+		SendType string
+		// Auth arguments loaded for the plugin
+		Auth string
+		// Env arguments loaded for the plugin
+		Env map[string]string
+		// Readable build time environment variables
+		BuildEnv *BuildEnv
+	}
+
+	// SMTPHost struct.
+	SMTPHost struct {
+		Host     string
+		Port     string
+		Username string
+		Password string
+	}
+
+	// User friendly readable Build Environment Variables.
+	BuildEnv struct {
+		BuildCreated  string
+		BuildEnqueued string
+		BuildFinished string
+		BuildStarted  string
+	}
+)
+
+// Validate checks the plugin parameters needed from the
+// user are provided. If the email subject or HTML/text are
+// not provided, defaults are set.
+func (p *Plugin) Validate() error {
+	logrus.Trace("entered plugin.Validate")
+	defer logrus.Trace("exited plugin.Validate")
+
+	logrus.Info("Validating Parameters...")
+	if len(p.Attachment.Filename) != 0 {
+		fileInfo, err := os.Stat(p.Attachment.Filename)
+		if errors.Is(err, os.ErrNotExist) {
+			return os.ErrNotExist
+		}
+
+		if fileInfo.Size() == 0 {
+			return ErrorEmptyAttach
+		}
+
+		// Don't check email's errors
+		file, err := os.Open(p.Attachment.Filename)
+		if err != nil {
+			return err
+		}
+
+		// Don't check email's errors
+		p.Email, err = email.NewEmailFromReader(file)
+		if err != nil {
+			return err
+		}
+
+		if len(p.Email.To) > 0 {
+			p.Email.To = stringToSlice(p.Email.To)
+		}
+
+		if len(p.Email.Cc) > 0 {
+			p.Email.Cc = stringToSlice(p.Email.Cc)
+		}
+
+		if len(p.Email.Bcc) > 0 {
+			p.Email.Bcc = stringToSlice(p.Email.Bcc)
+		}
+	}
+
+	if len(p.Email.To) == 0 {
+		return ErrorMissingEmailToParam
+	}
+
+	if len(p.Email.From) == 0 {
+		return ErrorMissingEmailFromParam
+	}
+
+	if len(p.SMTPHost.Host) == 0 || len(p.SMTPHost.Port) == 0 {
+		return ErrorMissingSMTPParam
+	}
+
+	if len(p.SMTPHost.Username) == 0 {
+		return ErrorMissingSMTPUsernameParam
+	}
+
+	if len(p.SMTPHost.Password) == 0 {
+		return ErrorMissingSMTPPasswordParam
+	}
```

Comment:
Can these be set as optional values instead? I have a use-case where I want to send emails through a smtp server that does require or support authentication.

Reply:
Which sendtype would you need to use? each of our sendtypes require authentication

Reply:
I'm not familiar with the github.com/jordan-wright/email package you are using, but the code comments make it seem like SMTP auth is optional for some of the functions - https://github.com/jordan-wright/email/blob/master/email.go#L507. I have also used gopkg.in/mail.v2 in the past to send emails without authentication by doing something like:

```go
m := mail.NewMessage()
m.SetHeader("From", cfg.EmailFrom)
m.SetHeader("To", email)
m.SetHeader("Subject", cfg.EmailSubject)
m.SetBody("text/html", body)

d := mail.Dialer{Host: cfg.SMTPRelay, Port: 25}
if err := d.DialAndSend(m); err != nil {
  return err
}
```